### DISCUSSION TITLE
dynamic component

### DISCUSSION THREAD

**Discussion by flosigime** [sqlpage/SQLPage#223] [Q&A]
Hello everybody, 

I'm trying to make a dynamic shell, that will have a dynamic title based on data in our database, and a footer containing some links and even the sqlpage version with the sqlpage.version() function. 

It works fine if i write it on each page, but i'm wandering if i could use the dynamic component to create the shell, not to have to change the shell on each page, like an include in php. 

I made some tries, using the examples in the documentation, but it seem that the only way is to use the sqlpage.read_file_as_text function, which is absolutely static and can't use sql request in it ? 
Am i right ? 
An idea of how to do ? 
Thank you

**Comment by lovasoa** [comment]
Hello !

  There is currently no "inclusion" mechanism in SQLPage. The feature has been [asked before](https://github.com/lovasoa/SQLpage/discussions/42) and [will be added to a later version](https://github.com/lovasoa/SQLpage/issues/20).

  However, if you can do without the call to `sqlpage.version()`, a concise and elegant solution to your problem exists already: **define your shell in a SQL view**. This is a pure sql solution, that does not require any special SQLPage feature:

#### `sqlpage/migrations/0001_define_shell.sql`
```sql
create view sqlpage_shell as
select
    'shell' as component,
    'My Website' as title,
    json_group_array(category)  -- this is an example
from website_sections;
```
#### `index.sql`
```sql
select * from sqlpage_shell;
```

([try online](https://www.db-fiddle.com/f/k5P3avzo21LcAX31xRQJGq/0))

**Comment by flosigime** [comment]
Hello,

Yes, it's a really interesting way to do.
I really have to study migrations...


Le lun. 12 févr. 2024, 10:14, Ophir LOJKINE ***@***.***> a
écrit :

**Reply by lovasoa** [reply]
If you don't want to write a migration, you can create the view using whatever tool you use to administer your database.
